### ISSUE TITLE
How can I check if the layer I added is right or wrong?

### ISSUE THREAD

**Issue by ginobilinie** [BVLC/caffe#3654]
Hi,

Recently I add some layers to caffe, and passed the test in caffe. 

While when I try to train the model involving the new added layers, I found the loss is very hard decrease and the accuracy in test dataset is also not improved. I'm wondering if this means there is something wrong with the layers I add? And How can I make sure which parts of the added layer is wrong?

Many thanks.

**Comment by seanbell** [comment]
Write a unit test.   Look at the other unit tests for examples of how to write them.

**Comment by ginobilinie** [comment]
@seanbell 
Thank you. I have write a unit test following the examples. And my added layers all passed the unit test. 
While when I use the added layer to train a model, I found the loss is decreasing, but the accuracy for test dataset doesn't increase. So I'm wondering if my added layers are right or not.

**Comment by ShaggO** [comment]
If your unit tests are correctly written and your implementation passes them, the layers should be implemented correctly (the way you've defined them, that is).
The loss reported by Caffe is only calculated on the basis of the batch you've processed (as far as I remember) and not the entire training set, whereas the test accuracy is computed as defined by the test parameters ("batch_size" for phase: test and "test_iter"). A decreasing train loss and increasing test accuracy could have numerous causes: a small train or test batch size and/or incorrect handling/preprocessing of the dataset ( shuffling in example), an "incomplete" dataset (not covering the complete variation of the data), a bad split between train and test data, or overfitting of parameters.
I believe these issues are usage-related (and should be discussed on the mailing list).

**Comment by ginobilinie** [comment]
@ShaggO 
Thank you so much. What you write is helpful a lot. I'll check the whole process again.

**Comment by ShaggO** [comment]
@ginobilinie you're welcome. I suppose we can close this issue now?

**Comment by ginobilinie** [comment]
@ShaggO  Of course. Thanks a lot.
